### PR TITLE
WT-12713 Generate set_oldest_timestamp operations

### PR DESCRIPTION
**Pull request by pmacko86**
This PR:
* Adds the oldest timestamp to `model::kv_database`.
* Adds the `set_oldest_timestamp` workload operation.
* Adds the corresponding support to the workload generator.
* Adds timestamp verification to tests and to the debug log verifier.
* Fixes a bug in the debug log parser, so that it can handle empty prepared transactions.
* Fixes a bug in the debug log based verifier, so that it better handles the previous key-value pair when printing an error.

This ended up being a surprisingly substantial PR; feel free to let me know if you'd like a walk-through.

### CODE REVIEW

File: test/model/src/include/model/kv_database.h
Comment (korteland) [inline test/model/src/include/model/kv_database.h:116]: I'm not 100% sure that this matches WT behaviour - isn't it an error to try to make the oldest timestamp go backwards?

**Review by korteland — COMMENTED** [review]
Mostly LGTM - the details largely look fine but a quick chat to see how it hangs together overall and get the bigger picture would be useful to me.

File: test/model/src/core/kv_database.cpp
Comment (etienneptl) [inline test/model/src/core/kv_database.cpp:105]: That's an interesting case, is that what WT does as well?

File: test/model/src/driver/debug_log_parser.cpp
Comment (etienneptl) [inline test/model/src/driver/debug_log_parser.cpp:564]: np: This comment might be unexpected after the comment line 562. Maybe we could merge them both line 562?

File: test/model/src/driver/kv_workload_generator.cpp
Comment (etienneptl) [inline test/model/src/driver/kv_workload_generator.cpp:219]: I have a question regarding lines above which manage the stable ts.
What happens in this scenario:

- Start WT, not ts are set
- Set oldest to 10
- Set stable to 9

The above will not trigger as we set stable for the first time, but it is lower than the oldest ts.

File: test/model/src/include/model/kv_database.h
Comment (etienneptl) [reply test/model/src/include/model/kv_database.h:116]: It is an error, as described [here](https://github.com/wiredtiger/wiredtiger/pull/10427/files#diff-ba4c9e2fbaeaefdd722d5c7566e096e249c1ad3b59214e6b7443061ea9e9be2eR223).
Taking the max here means we should never have a lower value than the current `_oldest_timestamp`, it should be fine.

File: test/model/test/common/wiredtiger_util.cpp
Comment (etienneptl) [reply test/model/test/common/wiredtiger_util.cpp:92]: Same question, shouldn't we fail for an invalid ts?

File: test/model/test/common/wiredtiger_util.cpp
Comment (etienneptl) [inline test/model/test/common/wiredtiger_util.cpp:428]: Can we create private helper functions for `wt_[get|set]_[stable|oldest]_timestamp`? For example:
```
wt_set_timestamp(WT_CONNECTION *conn, const std::string& timestamp_name, model::timestamp_t timestamp)
{
    char buf[64];

    testutil_snprintf(buf, sizeof(buf), "%s=%" PRIx64, timestamp_name, timestamp);
    testutil_check(conn->set_timestamp(conn, buf));
}
```
What do you think?

**Review by etienneptl — COMMENTED** [review]
Added a few comments, I will need another review.

File: test/model/src/core/kv_database.cpp
Comment (pmacko86) [reply test/model/src/core/kv_database.cpp:105]: Yes. I don't know if it's documented, but empirically I see that this is what happens. You can check that the unit tests verify that WT has the same behavior. But I agree that it is somewhat unexpected!

File: test/model/src/include/model/kv_database.h
Comment (pmacko86) [reply test/model/src/include/model/kv_database.h:116]: When trying this empirically, this is the same behavior as I get with WT, and I added unit tests to make sure.

It is not currently an error for the oldest timestamp to be set to an earlier value, just like with the stable timestamp (at least until Etienne's PR gets merged in). I still didn't want to generate that behavior in the workload generator, which is why I added the check.

File: test/model/test/common/wiredtiger_util.cpp
Comment (pmacko86) [reply test/model/test/common/wiredtiger_util.cpp:92]: I need this for the unit tests that check that the model has the same behavior as WT when using a too-old read timestamp. Without this, doing the test on WT would result in an abort, but now I can just compare the return values.

Comment (korteland) [reply test/model/test/common/wiredtiger_util.cpp:92]: Got it - thanks.

**Review by korteland — APPROVED** [review]
LGTM! Thanks for the walkthrough.

File: test/model/src/driver/debug_log_parser.cpp
Comment (pmacko86) [reply test/model/src/driver/debug_log_parser.cpp]: I rewrote the comment in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

File: test/model/src/driver/kv_workload_generator.cpp
Comment (pmacko86) [reply test/model/src/driver/kv_workload_generator.cpp:225]: I also added a similar check to the stable timestamp branch above in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

File: test/model/test/common/wiredtiger_util.cpp
Comment (pmacko86) [reply test/model/test/common/wiredtiger_util.cpp:428]: Good idea—fixed in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

**Comment by pmacko86** [comment]
In addition to addressing the PR comments, I also discovered and fixed a bug in which the workload generator would not properly keep track of the oldest and stable timestamps across database crashes, which would result in generating workloads like this:

- Set stable timestamp to 5
- Checkpoint
- Set oldest timestamp to 10
- Set stable timestamp to 20
- Crash
- Set oldest timestamp to 15 (thinking that the stable timestamp is still 20)

This is also included in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

File: test/model/src/core/kv_database.cpp
Comment (etienneptl) [reply test/model/src/core/kv_database.cpp:105]: It has been confirmed that this is how WT works. Whether it is the right behaviour or not, it is being discussed within the team. The outcome is beyond the scope of this PR - resolving.

**Review by etienneptl — APPROVED** [review]
LGTM!

**Pull request merged**
